### DISCUSSION TITLE
SPI MOSI not rising to full voltage levels

### DISCUSSION THREAD

**Discussion by Gabe-From-Onda** [SpenceKonde/megaTinyCore#640] [Q&A] [answered]
Hi. I'm trying to get an attiny1626 to communicate with a 2" 320x240 LCD screen module with an ST7789VW driver using adafruit's ADAFRUIT_GFX library, referencing this [tutorial](https://learn.adafruit.com/2-0-inch-320-x-240-color-ips-tft-display?view=all). 

When I got everything hooked up, nothing happened except the screen's backlight turned on. So I hooked the MOSI and SCK to an oscilloscope and got this while the screen was still attached.
MOSI: CH1 (yellow)
SCK: CH2 (blue)
![TINY SPI0](https://user-images.githubusercontent.com/98503154/152268114-6f78214d-d532-48b4-bed9-590e1eed181e.jpg)
I'm not sure why the MOSI only rises to 1.8V sometimes, then rises to 5V later.

When I disconnect the screen module, the waveform seems fine:
![TINY SPI WO SCREEN0](https://user-images.githubusercontent.com/98503154/152268725-b7382648-faf1-4967-a9ca-d7bee247f98d.jpg)

And when I have an arduino nano with similar code(io pins changed to reflect different microprocessor) it can control the screen just fine without the stepping MOSI waveform:
![NANO SPI0](https://user-images.githubusercontent.com/98503154/152268867-6bc6aef9-42a2-4c2f-972d-a6df5e7e0031.jpg)

Any idea what could be causing the MOSI to pause at 1.8V when the screen is connected, but work normally when it's disconnected?

**Comment by SpenceKonde** [comment]
Could the power and clock pins be reversed? That looks to me like it's hitting a protection diode and that's keeping it from risink any further - yet jonce the clock goes high all of a sudden the data lne can too? 

That's an electrical problem IMO, whatever it is

**Comment by Gabe-From-Onda** [comment, marked as answer]
It ended up being an electrical issue. I think the clock and data lines were rising too fast and causing the VCC to be unstable. Once I put some resistors between the I/O pins and the screen, to slow the rise/fall time, everything started working.

**Comment by SpenceKonde** [comment]
Pretty sure it was actually not that. 

If I had been paying better attention the first time I'd have spotted what is jumping out at me now.... When you see data lines that should be digital taking intermediate values, but not over a continuous spectrum, rather discrete voltages, typically right in the middle, or at other integer fractions of what it should be.... and shifting between those voltagesin response to other data lines transitioning.... that is the calling card of grounds being connected only through a data line or other high impedance path (such as "through a data line by way of the protection diodes on one side or the other"). And it's also exactly the kind of error that frequently fixes itself when you are fiddling with the device, because it is most often a mechanical defect impacting a connector, and it could easily look like an unstable supply voltage depending on how it was being measured, and explain your claimed reason. 

Usually the bad connector in question is a "dupont" connector (ie, chinese knockoff of harwin's knockoff of the DuPont connector) because those things are both ubiquitous and terribly unreliable (the knockoffs I mean - the real ones are great, but they cost > 20 cents per terminal),

**Reply by Gabe-From-Onda** [reply]
Sounds reasonable (this is a very cheap Chinese screen) but I have a little bit of doubt because I switched between using 2 different identical screens, with their own cables on both the Arduino nano and the attiny, and the results were consistent after multiple tests. Once the resistors were installed, the attiny started working perfectly reliably.

**Comment by cj190133** [comment]
Hi Gabe-From-Onda,
for me it seems display is 3V device and ATTINY is powered with 5V...  And the signal from MOSI in high level opens protect diode in display pin ...

**Reply by Gabe-From-Onda** [reply]
The display isn't the exact module that Adafruit sells, but a cheap Chinese knockoff that's advertised as 5V tolerant for power and all I/O pins. Looking back, I'm not sure if I made that clear in the original post. It uses the same ST7789 driver and 320 x 240 resolution so it's identical as far as the code is concerned. There also isn't any significant voltage drop between the attiny I/O and the LCD module's I/O, with the 1kΩ resistors installed between them. So I'm a little bit skeptical that it's a voltage protection system.

**Reply by cj190133** [reply]
@Gabe-From-Onda 
the output voltage from ATtiny is significantly degreased - think about it - it is possible only when the output pin is overloaded ... It is electrical problem, hard to say the reason, when we don't see the schema ...
You said, if between Attiny MOSI pin and display is 1 kOhm resistor, voltage level is OK. It means, with resistor, the current is limited to less than 5 mA, therefore the voltage level corresponds to datasheet.
The display pin overloads the pin from Attiny ... there can be more reasons ... one of them is also defected display.

**Reply by Gabe-From-Onda** [reply]
It could be a current overload from some sort of capacitive load. The attiny has a 40mA current rating per pin, 100mA per pin group, and 200mA through VCC and GND, which is nothing to scoff at when you compare it to the 5mA that some other microprocessors give(which still handle SPI). There are probably other ways to solve this. Maybe with a buffer/redriver. A better LCD module probably wouldn't hurt. But the 1kΩ resistors were cheap and already on the BOM. 

Also, I tested it with several screens from the manufacturer and got the same results between several screens. So I don't think they're necessarily "defective" but probably just cheap.

**Reply by SpenceKonde** [reply]
A series resistor and reliance on protection diodes of the lower voltage device is extremely common.
